### PR TITLE
[narwhal] additional consensus tests

### PR DESCRIPTION
**Pull request by akichidis**
* small refactoring in the consensus behaviour when trying to insert a certificate of round lower that the last committed for the authority
* additional tests

### CODE REVIEW

**Review by laura-makdah — APPROVED** [review]

**Comment by mwtian** [comment]
I like the added tests but not sure I agree with the extra error statuses. With errors, usually people would want to log them or propagate upward. I think we need to differentiate a bit between
- Expected outcome from normal operations
- Errors that stop the operation

In this case, I think cert equivocation should be prevented by Narwhal. Other cases e.g. ignoring cert below commit round or with the same digest do not seem like real errors. But I'm ok with differentiating the cases as long as it is clear.

**Comment by akichidis** [comment]
> I like the added tests but not sure I agree with the extra error statuses. With errors, usually people would want to log them or propagate upward. I think we need to differentiate a bit between
> 
> * Expected outcome from normal operations
> * Errors that stop the operation
> 
> In this case, I think cert equivocation should be prevented by Narwhal. Other cases e.g. ignoring cert below commit round or with the same digest do not seem like real errors. But I'm ok with differentiating the cases as long as it is clear.

@mwtian I was not 100% sure about converting those to errors either, but basically wanted to get some strong signal that the operation we try to do does not end up in the DAG. Overall my thinking is that DAG should try to be as robust as possible on a component on its own and do some extra checks as well - basically some extra checks and balances. That being said, I could probably remove the `CertificatePassedCommit` error (leave the equivocation protection there) and maybe return as part of the `OK` result and enum with the operation outcome?

**Comment by akichidis** [comment]
@mwtian I've updated the solution to return an enum with the result action. Let me know if that looks better.

File: narwhal/consensus/src/consensus.rs [hunk 1]
```diff
@@ -97,7 +97,7 @@ impl ConsensusState {
 
         let mut num_certs = 0;
         for cert in &certificates {
-            if Self::try_insert_in_dag(&mut dag, last_committed, cert) {
+            if Self::try_insert_in_dag(&mut dag, last_committed, cert).is_ok() {
```
Comment (mwtian) [inline narwhal/consensus/src/consensus.rs:100]: We should check for Ok(true) here.

File: narwhal/consensus/src/consensus.rs [hunk 2]
```diff
@@ -131,16 +131,22 @@ impl ConsensusState {
...
-                certificate.origin(),
-                (certificate.digest(), certificate.clone()),
-            )
-            .is_none()
+        if let Some((_, existing_certificate)) = dag.entry(certificate.round()).or_default().insert(
+            certificate.origin(),
+            (certificate.digest(), certificate.clone()),
+        ) {
+            // we want to error only if we try to insert a different certificate in the dag
+            if existing_certificate != certificate.clone() {
```
Comment (mwtian) [inline narwhal/consensus/src/consensus.rs:142]: Can we just compare digest?

**Review by mwtian — APPROVED** [review]

**Comment by mwtian** [comment]
I like the enum outcome approach!

File: narwhal/consensus/src/consensus.rs (hunk 1 above)
Comment (akichidis) [reply narwhal/consensus/src/consensus.rs:100]: Oh my, yes, thanks @mwtian !

File: narwhal/consensus/src/consensus.rs (hunk 2 above)
Comment (akichidis) [reply narwhal/consensus/src/consensus.rs:142]: We could , I just used certificate comparison as it's implementing the `Eq` trait . But sure we can compare digests

**Pull request merged**